File: packages/naver-map/ios/RNNaverMapViewManager.m
//
//  RNNaverMapViewManager.m
//  react-native-toss-naver-map
//

#import <React/RCTBridge.h>
#import <React/RCTViewManager.h>
#import "RCTConvert+NMFMapView.h"

@interface RCT_EXTERN_MODULE(RNNaverMapViewManager, RCTViewManager)

RCT_CUSTOM_VIEW_PROPERTY(center, NMFCameraUpdateWith*, RNNaverMapViewManager)
{
    if (json == nil) return;
    [self performSelector: NSSelectorFromString(@"moveCamera:view:")
               withObject:[RCTConvert NMFCameraUpdateWith: json]
               withObject: view];
}

RCT_EXPORT_VIEW_PROPERTY(showsMyLocationButton, BOOL)
RCT_EXPORT_VIEW_PROPERTY(mapPadding, UIEdgeInsets)
RCT_EXPORT_VIEW_PROPERTY(mapType, int)
RCT_EXPORT_VIEW_PROPERTY(compass, BOOL)
RCT_EXPORT_VIEW_PROPERTY(scaleBar, BOOL)
RCT_EXPORT_VIEW_PROPERTY(zoomControl, BOOL)
RCT_EXPORT_VIEW_PROPERTY(buildingHeight, float)
RCT_EXPORT_VIEW_PROPERTY(nightMode, BOOL)
RCT_EXPORT_VIEW_PROPERTY(minZoomLevel, double)
RCT_EXPORT_VIEW_PROPERTY(maxZoomLevel, double)
RCT_EXPORT_VIEW_PROPERTY(scrollGesturesEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(zoomGesturesEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(tiltGesturesEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(rotateGesturesEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(stopGesturesEnabled, BOOL)
RCT_EXPORT_VIEW_PROPERTY(tilt, BOOL)
RCT_EXPORT_VIEW_PROPERTY(onInitialized, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCameraChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onTouch, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMapClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(locationTrackingMode, int)

RCT_EXTERN_METHOD(setLayerGroupEnabled:(nonnull NSNumber *)reactTag
                  withGroup:(NSString *)group
                  withEnabled:(BOOL)enabled)

RCT_EXTERN_METHOD(animateToCoordinate:(nonnull NSNumber *)reactTag
                  withCoord:(NMGLatLng *)coord)

RCT_EXTERN_METHOD(animateToTwoCoordinates:(nonnull NSNumber *)reactTag
                  withCoord1:(NMGLatLng *)coord1
                  withCoord2:(NMGLatLng *)coord2)

RCT_EXTERN_METHOD(animateToRegion:(nonnull NSNumber *)reactTag
                  withBounds:(NMGLatLngBounds *)bounds)

RCT_EXTERN_METHOD(addMarker:(nonnull NSNumber *)reactTag
                  identifier:(NSString *)identifier
                  markerData:(NSDictionary *)markerData)

RCT_EXTERN_METHOD(updateMarker:(nonnull NSNumber *)reactTag
                  identifier:(NSString *)identifier
                  markerData:(NSDictionary *)markerData)

RCT_EXTERN_METHOD(removeMarker:(nonnull NSNumber *)reactTag
                  identifier:(NSString *)identifier)

@end

File: packages/naver-map/ios/RNNaverMapViewManager.swift
//
//  RNNaverMapViewManager.swift
//  react-native-toss-naver-map
//

import NMapsMap
import React
import UIKit

@objc(RNNaverMapViewManager)
class RNNaverMapViewManager: RCTViewManager {

    var markers: [String: NMFMarker] = [:]

    override func view() -> UIView! {
        let mapView = RNNaverMapView(frame: CGRect(x: 0, y: 0, width: 200, height: 300))
        mapView.bridge = bridge
        return mapView
    }

    override static func requiresMainQueueSetup() -> Bool {
        false
    }

    @objc static func performActionWithView(
        reactTag: NSNumber,
        bridge: RCTBridge,
        action: @escaping (RNNaverMapView) -> Void
    ) {
        bridge.uiManager.addUIBlock { (_, viewRegistry) in
            guard let view = viewRegistry?[reactTag] as? RNNaverMapView else {
//                RCTLogError("Invalid view returned from registry, expecting RNNaverMapView, tag: \(reactTag)")
                return
            }
            action(view)
        }
    }

    static let NMFLayerGroup: [String: String] = [
        "building": NMF_LAYER_GROUP_BUILDING,
        "ctt": NMF_LAYER_GROUP_TRAFFIC,
        "transit": NMF_LAYER_GROUP_TRANSIT,
        "bike": NMF_LAYER_GROUP_BICYCLE,
        "mountain": NMF_LAYER_GROUP_MOUNTAIN,
        "landparcel": NMF_LAYER_GROUP_CADASTRAL
    ]

    @objc(setLayerGroupEnabled:withGroup:withEnabled:)
    func setLayerGroupEnabled(_ reactTag: NSNumber, withGroup group: String, withEnabled enabled: Bool) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { view in
            guard let layerGroup = Self.NMFLayerGroup[group] else {
                return
            }
            view.mapView.setLayerGroup(layerGroup, isEnabled: enabled)
        }
    }

    @objc(animateToCoordinate:withCoord:)
    func animateToCoordinate(_ reactTag: NSNumber, withCoord coord: NMGLatLng) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { view in
            let cameraUpdate = NMFCameraUpdate(scrollTo: coord)
            cameraUpdate.animation = .easeIn
            view.mapView.moveCamera(cameraUpdate)
        }
    }

    @objc(animateToTwoCoordinates:withCoord1:withCoord2:)
    func animateToTwoCoordinates(_ reactTag: NSNumber, withCoord1 coord1: NMGLatLng, withCoord2 coord2: NMGLatLng) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { view in
            let bounds = NMGLatLngBounds(
                southWestLat: max(coord1.lat, coord2.lat),
                southWestLng: min(coord1.lng, coord2.lng),
                northEastLat: min(coord1.lat, coord2.lat),
                northEastLng: max(coord1.lng, coord2.lng)
            )
            let cameraUpdate = NMFCameraUpdate(fit: bounds, padding: 24.0)
            cameraUpdate.animation = .easeIn
            view.mapView.moveCamera(cameraUpdate)
        }
    }

    @objc(animateToRegion:withBounds:)
    func animateToRegion(_ reactTag: NSNumber, withBounds bounds: NMGLatLngBounds) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { view in
            let cameraUpdate = NMFCameraUpdate(fit: bounds, padding: 0.0)
            cameraUpdate.animation = .easeIn
            view.mapView.moveCamera(cameraUpdate)
        }
    }

    @objc(addMarker:identifier:markerData:)
    func addMarker(_ reactTag: NSNumber, identifier: String, markerData: [String: Any]) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { [weak self] view in
            guard let self else { return }
            let mapView = view.mapView
            let data = NaverMapMarkerData(object: markerData)
            let marker = NMFMarker()
            marker.applyMarkerData(data)
            marker.mapView = mapView
            self.markers[identifier] = marker

            marker.touchHandler = { _ -> Bool in
                view.onMarkerClick?(["id": identifier])
                return true
            }
        }
    }

    @objc(updateMarker:identifier:markerData:)
    func updateMarker(_ reactTag: NSNumber, identifier: String, markerData: [String: Any]) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { [weak self] view in
            guard let self else { return }
            let mapView = view.mapView
            guard let marker = self.markers[identifier] else {
                return
            }
            let data = NaverMapMarkerData(object: markerData)
            marker.applyMarkerData(data)
            marker.mapView = mapView
        }
    }

    @objc(removeMarker:identifier:)
    func removeMarker(_ reactTag: NSNumber, identifier: String) {
        Self.performActionWithView(reactTag: reactTag, bridge: bridge) { [weak self] _ in
            guard let self else { return }
            guard let marker = self.markers[identifier] else {
                return
            }
            marker.mapView = nil
            self.markers[identifier] = nil
        }
    }

    @objc func moveCamera(_ update: NMFCameraUpdate, view: UIView) {
        guard let mapView = view as? RNNaverMapView else {
            return
        }
        Task { @MainActor [weak mapView] in
            guard let mapView = mapView?.mapView else { return }
            mapView.moveCamera(update)
        }
    }
}
